### DISCUSSION TITLE
Auth with Token

### DISCUSSION THREAD

**Discussion by thaarrtt** [pocketbase/pocketbase#4333] [Q&A]
is there a way to auth with token only ??? not auth refresh because refresh reset the token

**Comment by ganigeorgiev** [comment]
I don't understand your question. The authentication already happens only via token.

> not auth refresh because refresh reset the token

That is not true. Refresh just generates another token with refreshed `exp` claim.

**Reply by thaarrtt** [reply]
I want use pocketbase on another backend and wonder if auth refresh is right choice

**Reply by gokulgseven** [reply]
i used pocketbase with express in a particular project, 

How i went about it was that, i access pocketbase sqlite file with express directly (since both express and pocketbase were runing in the same docker)
with this i was able to create user specific tokens , with which i was able to use pocketbase apis , with the users authentication from express, (while the user specific api rules worked) (without having to save the users username and password in the database)

i would suggest go with the approach if you want to use the api rules , and only want to make a few apis, (this will be faster), if you want to use some of the apis provided by pocketbase and some of the apis that from your own backed.  (with this approach you can use the realtime apis of pocketbase , and the prebuilt apis of pocketbase , where ever required, and then use your own backend for apis that are there, basically you get the best of both worlds)

if you want to make more than a few apis , use the admin credentials , that gives you access to all the data , (you can still use api rules , based on request parameters , but not based on auth parameters. (this approach will be much simpler, but you will have to manage the auth flow in your app , and pocketbase will be just a database) 


also there is no refresh token flow , on pocketbase , as far as i am aware , maybe it changed in the newer versions

**Reply by danielo515** [reply]
I am very interested in knowing how you were able to issue tokens from your own server that were accepted by pocketbase without having to create users and without having to grant those users admin permissions.

**Reply by gokulgseven** [reply]
that is now what i did. (maybe i explained it wrong)

What i did was 
**Approach A**
1. created users in pocketbase 
2. created api rules for the users in pocketbase based on their id

3. created a common express based auth flow (in this directly connected to sqlite file to create user token)
this user token created by express auth flow was then saved to the local storage )
4. now i have a token in local storage that authenticates , my app in express and pocketbase (so can use api from either of the app with a single token), i can also create a token from either of the apps , 
5. used reverse proxy nginx to manage the routes for both backends . 

(how i used it :  i made majority of the things in pocketbase and its custom javascript apis , and then just added the AWS specific code in express , since i was not able to import the aws module inside pocketbase javascript, did my auth flow in express and generated a token , used this token with all pocketbase apis and a few express apis )

**Approach B**
1. your admin token that you create when you start pocketbase the first time , these credentials give you access to all databases , and apis , 
so basically you daisy chain the 2 backends together 
(not a good approach , )

**Reply by danielo515** [reply]
The point 3 of approach A is what is not clear to me. What are the required token claims ? Do you store it as plain text in the sqlite? In which table do you store the token?

**Reply by gokulgseven** [reply]
```
getToken: async (req, res) => {
        try {

            const userId = req.query.userId;

            const userResult = await getUser(userId);
            if (!userResult) {
                return res.status(404).json({ error: 'User not found' });
            }
            const secret = param.secret;
            const secretKey = userResult.tokenKey + secret;
            const payload = {
                userId: userResult.id,
                email: userResult.email,
                name: userResult.name
            }
            const token = jwt.sign(payload, secretKey, { expiresIn: '1h' });

            const obj = {
                userRecords: userResult,
                token: token
            }
            return res.status(404).json({ message: "success", result: obj });


        } catch (error) {
            console.error("Async Error==========>>", error);
            res.status(500).json({ error: 'Internal Server Error' });
        }
    },
```
    
    
the secret is in the sqlite file _params table ,( its in a single field its a large json , it contains all your app level details smtp details , standard emails that you send, and all the tokens for different functions,  etc )
secretKey - is in the users table 


there is a quite detailed documentation on the token claims in tokens section in the documentation

**Reply by danielo515** [reply]
That is perfect, thanks!

**Comment by VictorioBerra** [comment]
@danielo515 Could you extend with go and build a token endpoint? You could create a user that you would use as a "client" like in traditional OAuth2.

**Reply by danielo515** [reply]
I don't want pocketbase to be sending tokens directly to my clients, I want to build and sign them on the server, then send them to the client so they can use it against pocketbase.
Or you are suggesting the endpoint to communicate to a server?
What do you mean by `A user that you would ese as a client` ?

**Reply by VictorioBerra** [reply]
@danielo515 I guess i misunderstood.

> I don't want pocketbase to be sending tokens directly to my clients

Why not? A major perk of PB is its auth stuff.

**Comment by danielo515** [comment]
Because I already have a custom auth system in place and I don't want to
replace it with pocket base. My requirements are not the typical: I don't
want users to sign up themselves, I don't want to send them emails and I
don't want them to need to remember a password. I want to create users
myself and sign tokens for them, the rest is out of my need

**Reply by VictorioBerra** [reply]
@danielo515 Im not an expert on the PB internals but you might be able to extend with GoLang and write your own token validator and set the user auth context and all that. Maybe @dschissler can shed some light.

**Reply by gokulgseven** [reply]
i am doing something similar in pocket base (for another project)
in that i am using cookies for authentication (you can find a thread here detailing exactly how to set it up), 
when a user logs in or signs up (i send them a message on telegram and whatsapp with a link) when then click on the link they are redirected to the home page already logged in to the app. (no password , no otp, no email)

all of this in javascript (inside pocketbase)(i dont know go yet , planning to learn it. but not getting time)

the pocketbase javascript can do quite a lot of things

**Reply by VictorioBerra** [reply]
@gokulgseven can you share the JS code that handles the callback from telegram?

**Reply by gokulgseven** [reply]
i am using a third party api for telegram and whatsapp (the only functionality i have is to send message to a number)

```
const sendTelegramMessage = $http.send({
    url:     `${telegramapi}/api/v1/sendmessage`,
    method:  "POST",
    body:    JSON.stringify({"message": message, "recieverNo": phone}),
    headers: {"content-type": "application/json", api-key: telegramapikey},
})
```
i dont think this will help you 
if i didnt have the third party api , i would have made the telegram api myself in express and ran it as a different service and just called at the api from pocketbase (i had a bad experience connecting aws npm package with pocketbase earlier , so now dont really connect any npm package to pocketbase at all) (if you know how to write in go maybe you can use that to connect to telegram , but i cant help you with that)

**Reply by danielo515** [reply]
@gokulgseven while it is nice of you to share your method, I don't see how it is similar to what I want to achieve.
You are handling the authentication on pocketbase (I assume the link contains some code coming from PB) and then pocketbase sets the cookie, for the existing pocketbase user. I don't want to have to save my users in PB

**Comment by danielo515** [comment]
I only used GO for very small and simple stuff, I don't feel like
implementing this with the tight deadline I have now, but maybe for a
future project.
Do you think this can also be achieved with JS?

On Fri, Mar 8, 2024 at 4:20 PM Victorio Berra ***@***.***>
wrote:

> @danielo515 <https://github.com/danielo515> Im not an expert on the PB
> internals but you might be able to extend with GoLang and write your own
> token validator and set the user auth context and all that. Maybe
> @dschissler <https://github.com/dschissler> can shed some light.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pocketbase/pocketbase/discussions/4333#discussioncomment-8721307>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWOIQJKFFQ7GLOUHFXLYXHJNXAVCNFSM6AAAAABDEEXMU6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DOMRRGMYDO>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com

**Reply by VictorioBerra** [reply]
Im not sure of that but I know if you use PB as a Go package you get some really low level control that you cant get with JS. Hopefully one the pros can chime in.

I will say maybe PB is not the answer if you don't want to use its auth stuff. Its a very big part of PB. You might as well use some other frameworks or libraries that lay down APIs over Sqlite and stuff. Im sure there are tons.

**Reply by danielo515** [reply]
Yeah, you're probably right. I was more interested in the real-time DB and handling of file uploads, which were also big selling points of pocket base. I was not aware that the auth part was such all or nothing.

**Reply by VictorioBerra** [reply]
@danielo515 Crazy idea kinda:

https://pocketbase.io/docs/js-event-hooks/#onrecordbeforeauthwithpasswordrequest

What if you use that to validate your token from your external auth server?

Use the JS SDK as normal, but pass the token as the PW. The username does not matter:

```javascript
const pb = new PocketBase('https://pocketbase.io');
const authData = await pb.collection('users').authWithPassword('whatever', 'TOKEN');
```

Then in the event hook I linked, you try and find the user in your users collection, if they arent there, create a user with a random PW (this PW will never be used!). This will actually issue a token back to the PB JS SDK so its used as normal from that point on.

This way, your users get created in PB so that they can be linked up and all that with files or whatever and you get to use tokens issues from your token server. Im sure you could put the username or ID as a claim in your token (if its not the sub) and use that to create the PB users.

**Reply by danielo515** [reply]
That is a very good idea, and it seems that is not that crazy, since it is probably one of the intended usecases of that hook.
I am not sure if what you are suggesting is to run that snippet in the client or in the server. If you are suggesting to do it all in the server, and then use the returned authData to set a token cookie that makes sense. However, why can't I just create the user directly from the server at the time of login, use the password there to log them ind and then set both tokens (mine and pocketbase) in the cookie?
The problem I see with using the hook is the ability to decode the token. I think my current way of validating a token (jose) requires some node APIs that may not be available within the PB Javascript Engine

**Reply by VictorioBerra** [reply]
The snippet is the PB JS SDK which is intended to be used in the browser. Any time you see `new PocketBase('127.0.0.1:1234')` you need to automatically recognize that as client-side PB JS.

`onrecordbeforeauthwithpasswordrequest` is intended to be used on the server. Just use jose `jwtVerify` there to validate the token, and extract your claims.

**NO setting cookies.** Just verify the token on the server with `onrecordbeforeauthwithpasswordrequest`. [Create the user](https://pocketbase.io/docs/js-records/#create-new-record) if they don't exist using the claim data in your token. Otherwise, [find the user](https://pocketbase.io/docs/js-records/#fetch-single-record) you want.

From there you are done. PB will issue a token as a response to the client and the PB JS SDK is used as normal. You only use your own token to kick off the process.

**Reply by danielo515** [reply]
Ok, if JOSE works fine in the PB JS engine, then this is almost perfect. Will give it a try, thanks

**Comment by thaarrtt** [comment]
out of topic maybe

for the guys that want more granular open source authentication and user access management 
I can recommend it to [Zitadel](https://zitadel.com/) or [Casdoor](https://casdoor.org/)

**Reply by VictorioBerra** [reply]
I personally like https://logto.io/

**Reply by hamza-imran75** [reply]
Does this work with pocketbase authentication?

**Comment by VictorioBerra** [comment]
Ah that could be an issue I didn't consider. Since it's crypto libraries
who knows. Maybe you could learn just enough golang to use this?
https://github.com/go-jose/go-jose

**Comment by danielo515** [comment]
That is what I was trying to say to begin with :smile:

On Wed, Mar 13, 2024 at 1:42 PM Victorio Berra ***@***.***>
wrote:

> Ah that could be an issue I didn't consider. Since it's crypto libraries
> who knows. Maybe you could learn just enough golang to use this?
> https://github.com/go-jose/go-jose
>
> On Wed, Mar 13, 2024, 4:08 AM Daniel Rodríguez Rivero <
> ***@***.***> wrote:
>
> > Ok, if JOSE works fine in the PB JS engine, then this is almost perfect.
> > Will give it a try, thanks
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/pocketbase/pocketbase/discussions/4333#discussioncomment-8770039>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAWMN23L5K4METPKPUY2FATYYACO7AVCNFSM6AAAAABDEEXMU6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DONZQGAZTS>
>
> > .
> > You are receiving this because you commented.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pocketbase/pocketbase/discussions/4333#discussioncomment-8772679>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWOMOO4GZJMTNS5GF4DYYBCUTAVCNFSM6AAAAABDEEXMU6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DONZSGY3TS>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com

**Comment by ganigeorgiev** [comment]
I'm closing this discussion as it starting to become too off-topic and diverge too much from the original question and could cause confusion to other users who search only by the title.

If anyone has further questions, please consider opening a new discussion with clear description and details what actually you are trying to do.

**Comment by OskarPetersilie** [comment]
Hey even tough the discussion is closed i want to share the solution for my usecase:

**Problem**: In a docker compose stack i have a Pcoketbase instance and a frontend whcih auths via pocketbase. 
 
 Now furthermore the FE is talking to an api in the stack and passes the Auth header.
 
 The api should now than verify the token and try to auth or rather get the user objec tfrom token so the api can act like the user which requested.
 
 **Solution**: 
 
 Note: even tough we call await pb.collection("users").authRefresh(); the token does not get invalidated and the FE session is fine
 
```js
    pb.authStore.save(token, null);
    const user = await pb.collection("users").authRefresh();
```
Source: [Reddit User answer](https://www.reddit.com/r/pocketbase/comments/15upbkl/comment/ke7h4eo/)

And the full code for context

```js
import express from 'express';
import cors from 'cors';
import PocketBase from 'pocketbase';


const app = express();
const port = 3000;
const pocketbaseUrl = process.env.POCKETBASE_URL || 'http://pocketbase:8090';


const pb = new PocketBase(pocketbaseUrl);

app.use(cors());
app.use(express.json());

// Middleware to verify authentication
const authMiddleware = async (req, res, next) => {

  // expextes Authorization header in format "Bearer <your_token>"
  // eventually adjust this to your use case

  const token = req.headers.authorization?.split(' ')[1];
  if (!token) {
    return res.status(401).json({ error: 'No token provided' });
  }
  try {
    pb.authStore.save(token, null);  // sets the token in authStore
    if (pb.authStore.isValid) {
      const user = await pb.collection("users").authRefresh(); // uses authStore with token to refresh and tehrefore load the user record
      req.user = user.record // saves user to request so it can be used in every route
      next();
    } else {
      res.status(401).json({ error: 'Invalid token' });
    }
  } catch (error) {
    res.status(401).json({ error: 'Authentication failed' });
  }
};

// Get user's saved routes
app.get('/routes', authMiddleware, async (req, res) => {
  try {
    const routes = await pb.collection('routes').getList(1, 50, {
      filter: `user = "${req.user.id}"`
    });
    res.json(routes);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch routes' });
  }
});


app.listen(port, () => {
  console.log(`OSRM Proxy listening at http://localhost:${port}`);
});

```

**Reply by danielo515** [reply]
> Note: even tough we call await pb.collection("users").authRefresh(); the token does not get invalidated and the FE session is fine

That is fantastic, and I appreciate that you share your solution with the rest of us. But that is undocumented behavior, and I prefer to use something that the PB team is aware that people are using it and doesn't get fixed on some random minor update later.

**Reply by OskarPetersilie** [reply]
As of today it is actually documented. ( https://pocketbase.io/docs/authentication/#auth-token-verification )

Pocketbase is very fast evolving and there where a lot of breaking changes (especially in the "extending with js" part). So as soon as you use pocketbase you should consider that codebase can break quite frequently until it hits v1.0
